src/software/thunderscope/play/qlearninginfo_widget.py

```diff
@@ -0,0 +1,137 @@
+        :param buffer_size: The buffer size, set higher for smoother plots.
+                            Set lower for more realtime plots. Default is arbitrary
+        """
+        QWidget.__init__(self)
+
+        self.linear_q_func_info_buffer = ThreadSafeBuffer(buffer_size, LinearQFunctionInfo, False)
+        self.linear_q_func_infos = {}
+
+        self.linear_q_func_weights_table = QTableWidget(0, 0)
+        self.linear_q_func_weights_table.horizontalHeader().hide()
+        self.linear_q_func_weights_table.verticalHeader().hide()
+
+        self.linear_q_func_combo_box = QComboBox()
+        self.linear_q_func_combo_box.currentTextChanged.connect(
+            self.__display_linear_q_function_weights
+        )
+
+        self.save_linear_q_func_weights_button = QPushButton("Save Weights")
+        self.save_linear_q_func_weights_button.clicked.connect(
+            self.__save_linear_q_function_weights
+        )
+
+        self.vertical_layout = QVBoxLayout()
+        self.vertical_layout.addWidget(self.linear_q_func_combo_box)
+        self.vertical_layout.addWidget(self.linear_q_func_weights_table)
+        self.vertical_layout.addWidget(self.save_linear_q_func_weights_button)
+        self.setLayout(self.vertical_layout)
+
+    def refresh(self) -> None:
+        """Update the Q-learning info widget with new information"""
+
+        linear_q_func_info = self.linear_q_func_info_buffer.get(block=False, return_cached=False)
+
+        # Updating QTableWidget could be expensive, so we only update if there is new data
+        if linear_q_func_info is None:
+            return
+
+        linear_q_func_info_dict = MessageToDict(linear_q_func_info)
+        linear_q_func_name = linear_q_func_info_dict["name"]
+
+        # Add this LinearQFunction to combo box if it is new
+        if linear_q_func_name not in self.linear_q_func_infos:
+            self.linear_q_func_combo_box.addItem(linear_q_func_name)
+
+        self.linear_q_func_infos[linear_q_func_name] = linear_q_func_info_dict 
+
+        # Update QTableWidget if this LinearQFunction is currently selected in combo box
+        if self.linear_q_func_combo_box.currentText() == linear_q_func_name: 
+            self.__display_linear_q_function_weights(linear_q_func_name)
+
+    def __display_linear_q_function_weights(self, linear_q_func_name: str) -> None:
+        """Update the QTableWidget with the weights of the specified LinearQFunction
+
+        :param linear_q_func_name: the name identifying the LinearQFunction
+        """
+
+        if linear_q_func_name not in self.linear_q_func_infos:
+            return
+
+        linear_q_func_info = self.linear_q_func_infos[linear_q_func_name]
+
+        num_features = linear_q_func_info["numFeatures"]
+        num_actions = linear_q_func_info["numActions"]
+
+        # Setting table size dynamically
+        self.linear_q_func_weights_table.setColumnCount(num_features)
+        self.linear_q_func_weights_table.setRowCount(num_actions)
```

Comment:
it would be nice to have feature titles at the top of each column

Reply:
I'll look into doing this later if I have time. Not sure if there's a better solution than storing the feature and action names in each LinearQFunctionInfo message? Because that might inflate our replay file sizes